crates/astria-sequencer/src/app/mod.rs

```diff
@@ -1184,25 +1284,48 @@ impl App {
         update_mempool_after_finalization(&mut self.mempool, &self.state, self.recost_mempool)
             .await;
 
-        let post_transaction_execution_result: PostTransactionExecutionResult = self
+        let PostTransactionExecutionResult {
+            events,
+            tx_results,
+            validator_updates,
+            mut consensus_param_updates,
+        } = self
             .state
             .object_get(POST_TRANSACTION_EXECUTION_RESULT_KEY)
             .expect(
                 "post_transaction_execution_result must be present, as txs were already executed \
                  just now or during the proposal phase",
             );
 
+        self.update_consensus_params_if_upgrade_due(
+            finalize_block.height,
+            &mut consensus_param_updates,
+        )
+        .await
+        .wrap_err("failed to update consensus params")?;
```

Comment:
i think having the upgrades handled in two separate (and somewhat specific) functions may get confusing in the future. what do you think about having an `UpgradeComponent` with explicit `begin_block` and `end_block` hooks, and the logic for `execute_upgrade_if_due` would go in the `begin_block` hook and the logic for `update_consensus_params_if_upgrade_due` would go in the `end_block` hook. i think that would be clearer/more explicit as to when certain changes should happen.

Reply:
I like the general idea, but the two main drawbacks I can see so far are:
- there's an open ticket #1809 which is relevant.  I'm hoping that removes the `Component` trait altogether as it's needlessly restrictive
- `execute_upgrade_if_due` which would be done in `UpgradeComponent::begin_block` currently returns `Result<Option<Vec<ChangeHash>>>`.  Since `begin_block` restricts the return type to `Result<()>`, I'd need to work around that by e.g. dumping the result into the ephemeral store, which I really dislike.

What do you think about providing an `UpgradeHandler` which is more akin to `ProposalHandler`?

Reply:
good points, I like the idea of an `UpgradeHandler`

Reply:
Done in [d7c16e0](https://github.com/astriaorg/astria/pull/1917/commits/d7c16e0c2011231ff639d5762d40eb2009151f5d).